[wip/rfc][nn-module] Speedup up common case of no hooks

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130344
* #130285

This reduces Dynamo guard overhead with inlining builtin nn modules by 25%-30% because we can replace a bunch of empty OrderedDict guards with just a bool guard. 

Opening the PR for discussion if direction is outright wrong and I should not pursue it.

anijain2305 committed:
[wip/rfc][nn-module] Speedup up common case of no hooks

[ghstack-poisoned]

albanD reviewed (commented):
Going from 8 dict checks to 4 + 1bool is nice but sounds like an incremental win that might not justify the potential footgun of inconsistency.

btw what's the status on using dict observers? Would this make all of this irrelevant by having a single bool for all dictionaries we guard on?

anijain2305 committed:
Update on "[wip/rfc][nn-module] Speedup up common case of no hooks"


This reduces Dynamo guard overhead with inlining builtin nn modules by 25%-30% because we can replace a bunch of empty OrderedDict guards with just a bool guard. 

Opening the PR for discussion if direction is outright wrong and I should not pursue it.

[ghstack-poisoned]

anijain2305 commented:
@albanD Yes, I have been looking at dict versions or dict watchers from 3.12. 

Unfortunately they don't help in the common case. Common case is not having any hooks. For such cases, when Dynamo inlines builtin nn modules, we observe 4 accessors and 4 guards as shown below
![image](https://github.com/pytorch/pytorch/assets/13822661/e849b906-a44b-4a59-9910-802c111dfbd7)

These guards are inserted for every builtin nn module, exploding the number of guards. These guards are just checking that the each hook dict is empty. There isn't much we can do here with dict tags. Even reading the tag might be more expensive per dictionary.

This PR converts those 4 accessors + 4 DICT_LENGTH guards into 1 accessor + 1 EQUALS_MATCH guard. And the good news is that using dict versions on the nn module `__dict__` itself, we can skip the EQUALS_MATCH as well.

anijain2305 commented:
Closing the PR for now. For now, TorchDynamo has taken a path of skipping guards on hooks. This introduces soundness issue but in a very controlled manner (not like earlier guard_nn_modules). We can circle back to this PR if the interest arises.
